aptos-move/mvhashmap/src/versioned_aggregators.rs

```diff
@@ -427,50 +427,187 @@ impl VersionedAggregators {
     /// Upon commit Snapshot and Delta entries are all required to be replaced with Values.
     pub fn record_value(
         &self,
-        id: AggregatorID,
+        id: K,
         txn_idx: TxnIndex,
         value: AggregatorValue,
-        maybe_apply: Option<AggregatorApplyChange>,
+        maybe_apply: Option<AggregatorApplyChange<K>>,
     ) {
         self.values
             .get_mut(&id)
             .expect("VersionedValue for an (resolved) ID must already exist")
             .insert(txn_idx, AggregatorEntry::Value(value, maybe_apply));
     }
 
-    pub fn record_apply(&self, id: AggregatorID, txn_idx: TxnIndex, apply: AggregatorApplyChange) {
+    pub fn record_apply(&self, id: K, txn_idx: TxnIndex, apply: AggregatorApplyChange<K>) {
         self.values
             .get_mut(&id)
             .expect("VersionedValue for an (resolved) ID must already exist")
             .insert(txn_idx, AggregatorEntry::Apply(apply));
     }
 
-    pub fn mark_estimate(&self, id: AggregatorID, txn_idx: TxnIndex) {
+    pub fn mark_estimate(&self, id: K, txn_idx: TxnIndex) {
         self.values
             .get_mut(&id)
             .expect("VersionedValue for an (resolved) ID must already exist")
             .mark_estimate(txn_idx);
     }
 
-    pub fn remove(&self, id: AggregatorID, txn_idx: TxnIndex) {
+    pub fn remove(&self, id: K, txn_idx: TxnIndex) {
         self.values
             .get_mut(&id)
             .expect("VersionedValue for an (resolved) ID must already exist")
             .remove(txn_idx);
     }
 
-    pub fn update_committed_idx(&self, committed_idx: TxnIndex) {
-        self.next_idx_to_commit
-            .fetch_max(committed_idx + 1, Ordering::SeqCst);
+    /// Moves the commit index, and computes exact values for aggregators having
+    /// apply changes in this transaction. After it finishes, all versions at or
+    /// before given idx are in Value state.
+    ///
+    /// Must be called for each transaction index, in order.
+    pub fn try_commit(&self, idx_to_commit: TxnIndex, ids: Vec<K>) -> Result<(), CommitError> {
+        // we may not need to return values here, we can just read them.
+        use AggregatorApplyChange::*;
+
+        if idx_to_commit != self.next_idx_to_commit.load(Ordering::SeqCst) {
```

Comment:
Just curious. If we want the try_commit to be executed only on `self.next_idx_to_commit`, why does this index have to be give as an input parameter to try_commit again?

Reply:
Caller knows which transaction it is trying to commit. Imagine there is a big, and they are off, and then caller calls, thinking it is committing X, and function succeeds in committing  Y , and caller thinks all is good